### PR TITLE
statem_srvr.c : Release pkey_ctx on initialization failure

### LINKED ISSUE #24341: There is a memory leak defect at line 3232 of the file /openssl/ssl/statem/statem_srvr.c.
At line 3197 of the file /openssl/ssl/statem/statem_srvr.c, a pointer variable named pkey_ctx is defined. This variable allocates a dynamic memory region through the function EVP_PKEY_CTX_new_from_pkey at line 3225. When the if statement at line 3226 returns false, it indicates that the dynamic memory region pointed to by pkey_ctx has been successfully allocated. However, if the if statement at line 3230 returns true, the program will return at line 3232 without executing the release operation for the dynamic memory region pointed to by pkey_ctx at line 3287. This leads to a memory leak defect, as shown in the diagram below:
https://github.com/LuMingYinDetect/openssl_defects/blob/main/openssl_22.png

### PR DESCRIPTION
**Pull request by irosay**
Fixes #24341

### CODE REVIEW

**Comment by t8m** [comment]
@irosay This would be ok with CLA: trivial. Could you please amend the commit message with `git commit --amend; git push --force` to include `CLA: trivial` in the commit message body on a separate line?

File: ssl/statem/statem_srvr.c [hunk 1]
```diff
@@ -3229,6 +3229,7 @@ static int tls_process_cke_gost(SSL_CONNECTION *s, PACKET *pkt)
         return 0;
     }
     if (EVP_PKEY_decrypt_init(pkey_ctx) <= 0) {
+        EVP_PKEY_CTX_free(pkey_ctx);
         SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
         return 0;
```
Comment (mattcaswell) [inline ssl/statem/statem_srvr.c:3234]: A fix more consistent with the rest of this function might be to instead replace this `return 0` with a `goto err`

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/statem/statem_srvr.c [hunk 2]
```diff
@@ -3229,8 +3229,9 @@ static int tls_process_cke_gost(SSL_CONNECTION *s, PACKET *pkt)
...
     }
     if (EVP_PKEY_decrypt_init(pkey_ctx) <= 0) {
+        EVP_PKEY_CTX_free(pkey_ctx);
```
Comment (mattcaswell) [inline ssl/statem/statem_srvr.c:3232]: This now isn't necessary since the err block already does it.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/statem/statem_srvr.c (hunk 2 above)
Comment (irosay) [reply ssl/statem/statem_srvr.c:3232]: okay thank you

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
Ok with trivial

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
OK with CLA: trivial

**Comment by t8m** [comment]
Merged to all the active branches. Thank you for your contribution.

**Pull request closed without merging**